### PR TITLE
search: factor out most functionality in searchFilesInRepos

### PR DESCRIPTION
**Pull request by keegancsmith**
This function has long since been confusing to use due to the size of it, and quite a few special cases around structural and searcher. This PR pulls out most of the logic into separate functions. This leads to a much easier to navigation function which is mainly responsible for calling out to zoekt/searcher/etc.

One of the main things done here is a switch to using errgroup more. Please review commit by commit.

### CODE REVIEW

File: cmd/frontend/graphqlbackend/textsearch.go [hunk 1]
```diff
@@ -650,6 +560,107 @@ func searchFilesInRepos(ctx context.Context, args *search.TextParameters, stream
...
+	// limiter of the new limit, which is a multiple of the number of
+	// searchers.
+	eps, err := args.SearcherURLs.Endpoints()
+	if err != nil {
+		return err
+	}
+	textSearchLimiter.SetLimit(len(eps) * 32)
+
+	g, ctx := errgroup.WithContext(ctx)
+	g.Go(func() error {
```
Comment (stefanhengl) [inline cmd/frontend/graphqlbackend/textsearch.go:527]: Is it really necessary to nest the loop kicking off goroutines within this goroutine? Just removing the first g.Go would simplify the code and lead to the same behavior, wouldn't it?

File: cmd/frontend/graphqlbackend/textsearch.go [hunk 2]
```diff
@@ -437,84 +429,44 @@ func searchFilesInRepos(ctx context.Context, args *search.TextParameters, stream
...
-		go func() {
-			defer wg.Done()
-			err := indexed.Search(ctx, stream)
-			if err != nil {
-				setError(ctx, stringerFunc("indexed"), err)
-				tr.LogFields(otlog.Error(err))
-			}
-		}()
+		g.Go(func() error {
+			return indexed.Search(ctx, stream)
```
Comment (stefanhengl) [inline cmd/frontend/graphqlbackend/textsearch.go:441]: It gets better and better!

**Review by stefanhengl — APPROVED** [review]
I am now officially a fan of errgroup!

**Comment by stefanhengl** [comment]
Just saw, StringerFunc is now unused, remove?

**Comment by rvantonder** [comment]
I kicked off a `master-dry-run` but our main CI is broken so now turns this PR red :( Will follow up and restart the build when main CI is fixed.

**Comment by rvantonder** [comment]
`master-dry-run` is green, thanks.

**Review by rvantonder — APPROVED** [review]
I think it's sensible to land this before ~#17928~ https://github.com/sourcegraph/sourcegraph/pull/17967 which may clash a little :-) I'll help rebase the other PR.

Also yay, this is great!! Thanks for pulling out `structuralSearchBacompt`, we will delete it promptly after branch cut, so that we have 1 iteration of backwards compatibility.

**Comment by keegancsmith** [comment]
> I kicked off a `master-dry-run` but our main CI is broken so now turns this PR red :( Will follow up and restart the build when main CI is fixed.
> 
> slack: https://sourcegraph.slack.com/archives/C07KZF47K/p1612375269203700

I already did that yesterday when I sent this PR out :P But thanks for double checking

**Comment by rvantonder** [comment]
> I already did that yesterday when I sent this PR out :P 

Ah  whoops, sorry! I wish there was an easy way to tell. I should have looked at the buildkite job first though.

**Pull request merged**